chain/client/src/stateless_validation/state_witness_tracker.rs

```diff
@@ -0,0 +1,213 @@
+use std::hash::{Hash, Hasher};
+use borsh::{BorshDeserialize, BorshSerialize};
+use lru::LruCache;
+use once_cell::sync::Lazy;
+
+use near_async::time::Clock;
+use near_o11y::metrics::{exponential_buckets, HistogramVec, try_create_histogram_vec};
+use near_primitives::sharding::ChunkHash;
+use near_primitives::stateless_validation::{ChunkStateWitness, ChunkStateWitnessAck};
+
+/// This is currently used for metrics computation, so we do not need to keep all the
+/// witnesses, so make the capacity 10.
+const CHUNK_STATE_WITNESS_MAX_RECORD_COUNT: usize = 5;
+
+static CHUNK_STATE_WITNESS_NETWORK_ROUNDTRIP_TIME: Lazy<HistogramVec> = Lazy::new(|| {
+    try_create_histogram_vec(
+        "near_chunk_state_witness_network_roundtrip_time",
+        "Time in seconds between sending state witness through the network to chunk producer \
+               and receiving the corresponding ack message",
+        &["witness_size_bucket"],
+        Some(exponential_buckets(0.01, 2.0, 12).unwrap()),
+    )
+        .unwrap()
+});
+
+/// Refers to a state witness sent to a chunk producer. Used to locate incoming acknowledgement
+/// messages back to the timing information of the originating witness.
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
+pub struct ChunkStateWitnessKey {
+    /// Hash of the chunk for which the state witness was generated.
+    pub chunk_hash: ChunkHash,
+}
+
+impl Hash for ChunkStateWitnessKey {
```

Comment:
you can derive hash implementation: `#[derive(Hash)]`

Reply:
done